erpnext/controllers/accounts_controller.py

```diff
@@ -965,67 +965,197 @@ def set_advance_gain_or_loss(self):
-						dr_or_cr = "debit" if dr_or_cr == "credit" else "credit"
+						if d.reference_doctype == "Purchase Invoice":
+							dr_or_cr = "debit" if dr_or_cr == "credit" else "credit"
 
-					gl_entries.append(
-						self.get_gl_dict(
+						reverse_dr_or_cr = "debit" if dr_or_cr == "credit" else "credit"
+
+						gain_loss_account = frappe.get_cached_value(
+							"Company", self.company, "exchange_gain_loss_account"
+						)
+						if not gain_loss_account:
+							frappe.throw(
+								_("Please set default Exchange Gain/Loss Account in Company {}").format(
+									self.get("company")
+								)
+							)
+						gain_loss_account_currency = get_account_currency(gain_loss_account)
+						if gain_loss_account_currency != self.company_currency:
+							frappe.throw(
+								_("Currency for {0} must be {1}").format(gain_loss_account, self.company_currency)
+							)
+
+						journal_account = frappe._dict(
 							{
-								"account": gain_loss_account,
-								"account_currency": account_currency,
-								"against": party,
-								dr_or_cr + "_in_account_currency": abs(d.exchange_gain_loss),
+								"account": party_account,
+								"party_type": self.party_type,
+								"party": self.party,
+								"account_currency": party_account_currency,
+								"exchange_rate": 0,
+								"cost_center": erpnext.get_default_cost_center(self.company),
+								"reference_type": d.reference_doctype,
+								"reference_name": d.reference_name,
+								"reference_detail_no": d.idx,
 								dr_or_cr: abs(d.exchange_gain_loss),
-								"cost_center": self.cost_center or erpnext.get_default_cost_center(self.company),
-								"project": self.project,
-							},
-							item=d,
+								dr_or_cr + "_in_account_currency": 0,
+							}
 						)
-					)
 
-					dr_or_cr = "debit" if dr_or_cr == "credit" else "credit"
+						journal_entry.append("accounts", journal_account)
 
-					gl_entries.append(
-						self.get_gl_dict(
+						journal_account = frappe._dict(
 							{
-								"account": party_account,
-								"party_type": party_type,
-								"party": party,
-								"against": gain_loss_account,
-								dr_or_cr + "_in_account_currency": flt(abs(d.exchange_gain_loss) / self.conversion_rate),
-								dr_or_cr: abs(d.exchange_gain_loss),
-								"cost_center": self.cost_center,
-								"project": self.project,
-							},
-							self.party_account_currency,
-							item=self,
+								"account": gain_loss_account,
+								"account_currency": gain_loss_account_currency,
+								"exchange_rate": 1,
+								"cost_center": erpnext.get_default_cost_center(self.company),
+								"reference_type": self.doctype,
+								"reference_name": self.name,
+								"reference_detail_no": d.idx,
+								reverse_dr_or_cr + "_in_account_currency": 0,
+								reverse_dr_or_cr: abs(d.exchange_gain_loss),
+							}
 						)
-					)
+
+						journal_entry.append("accounts", journal_account)
+
+						journal_entry.save()
+						journal_entry.submit()
```

Comment:
Show an alert to the user for Journal Entry creation. Also, add links in the Sales/Purchase and other documents dashboard (Connections tab) wherever applicable for the Journal Entry

Reply:
Added